core/src/main/java/com/scalar/db/storage/dynamo/QueryScanner.java

```diff
@@ -49,18 +58,23 @@ private boolean hasNext() {
       return true;
     }
     if (lastEvaluatedKey != null) {
-      handleResponse(request.execute(lastEvaluatedKey));
+      if (remainingLimit != null) {
+        handleResponse(request.execute(lastEvaluatedKey, remainingLimit));
+      } else {
+        handleResponse(request.execute(lastEvaluatedKey));
+      }
       return itemsIterator.hasNext();
     }
     return false;
   }
 
   private void handleResponse(PaginatedRequestResponse response) {
     List<Map<String, AttributeValue>> items = response.items();
-    totalResultCount += items.size();
+    if (remainingLimit != null) {
+      remainingLimit -= items.size();
+    }
     itemsIterator = items.iterator();
-    if ((request.limit() == null || totalResultCount < request.limit())
-        && response.hasLastEvaluatedKey()) {
+    if ((remainingLimit == null || remainingLimit > 0) && response.hasLastEvaluatedKey()) {
```

Comment:
If we have the remaining limit, we should send an additional request with the remaining limit to DynamoDB.